### DISCUSSION TITLE
[#Time Zone # pgSQL] Something wrong about Timestamp.

### DISCUSSION THREAD

**Discussion by kingwingfly** [vercel/community#3748] [Help] [answered]
### Summary

I'm a developer from China, so I got a problem about time zone: The query results (type is TIMESTAMP) of the same sql sentence differ between my computer's localhost and Vercel's deployment.

My computer is in China(UTC+8), and Vercel's service is in Washington, D.C., USA(UTC-4) which are in different time zones.

### Steps to Reproduce

I create a table by Vercel's postgreSQL service throught this:

```sql
CREATE TABLE meetings (
    id    SERIAL PRIMARY KEY,
    date_time    TIMESTAMP WITHOUT TIME ZONE
);
```
 Then I insert through this: 
```sql
INSERT INTO meetings (date_time)
VALUES ('2023-08-24 17:00:00');
```
Blow is what the table looks like in Vercel's web query console: 
|id|date_time|
|-|-|
|1|Thu Aug 24 2023 17:00:00 GMT+0800 (中国标准时间)|

Then I query though this:
```sql
SELECT date_time FROM meetings
WHERE id=1;
```

Things goes wrong then:
On my computer in China, it's:
```
date_time: 2023-08-24T09:00:00.000Z
```
On Vercel, the log shows:
```
date_time: 2023-08-24T17:00:00.000Z,
```
There are 8 hours distance always, how strange.

Here are my query code (looks like no big problem):
```ts
// ...
sql_ = `SELECT meeting_id, date_time, ... WHERE meeting_id IN (SELECT unnest(meetings) FROM users_meetings WHERE id = ${user_id});`;
console.log(sql_);
let meetings = (await client.query(sql_)).rows;
console.log(meetings);
// ...
``` 

This confused me a lot, why the results are different through the same sql sentence, and why it's 8 hours distance always?

It will be very kind of you for giving me a hand.

**Comment by amyegan** [comment]
Hey @kingwingfly. By default, [UTC](https://en.wikipedia.org/wiki/Coordinated_Universal_Time) is used for serverless functions and postgres regardless of the server's physical location. I think that accounts for the time difference.

**Reply by kingwingfly** [reply]
Thanks a lot!

I solved it for your telling me `TIMESTAMP WITHOUT TIME ZONE` has nothing to do with the location.

This occurs due to many reasons.

# First
This is what Vercel's **web** console return:

id | date_time
-- | --
1 | Thu Aug 24 2023 17:00:00 GMT+0800 (中国标准时间)

And it is **rendered** from the server through `(new Date('2023-08-24 17:00:00')).toString()`, which transformed according to the session between me and Vercel.

The real return is:
```
verceldb=> select id, date_time as no_tz from meetings;
 id |        no_tz
----+---------------------
  1 | 2023-08-24 17:00:00
```

# Second
Some implicit transforms happened.

`(await client.query(...)).rows` is of type **any[]**. This hides the fact that before `toISOString()` is called implicitly by `console.log()`, `rows[0].date_time` is actually already of type `Date` because `new Date(date_time_string)` has been called  implicitly.

The `date_time_string`(`2023-08-24 17:00:00`) having no time zone included, will be considered as a local time, and `toISOString()` during `console.log()` will lead to time zone transforming.

Here's an example:
```ts
// My computer is UTC+8

let t = new Date("2023-08-25 12:00:00"); // no time zone
console.log("without tz ISO:", t.toISOString());
// without tz ISO: 2023-08-25T04:00:00.000Z
console.log("without tz Local:", t.toLocaleString());
// without tz Local: 2023/8/25 12:00:00
t = new Date("2023-08-25 12:00:00+00"); // 'UTC' time zone
console.log("with tz ISO:", t.toISOString());
// with tz ISO: 2023-08-25T12:00:00.000Z
console.log("with tz Local:", t.toLocaleString());
// with tz Local: 2023/8/25 20:00:00
```

The two ISO outputs are exactly 8 hours away.

# Infer
Above all, we can infer:
1. Database's return is the same, which is UTC time but without timezone mark `+00`
2. On dev, it's considered as time_string without `+00`; On server, it's consider as time_string with `+00`.

# Verify
The data:
```
select * from meetings;
meeting_id |      date_time      
------------+--------------------
         16 | 2023-08-25 03:00:00
```
## On dev
Let's print the query result, `console.log(rows)`:
```
[
  {
    meeting_id: 16,
    date_time: 2023-08-24T19:00:00.000Z
  }
]
```
$24 - 19 + 3 = 8$，8 hours before 2023-08-25 03:00:00.
Fit in our guess.

## On server
Let's deploy it on Vercel and check the logs:
```
[
  {
    meeting_id: 16,
    date_time: 2023-08-25T03:00:00.000Z
  }
]
```
Difference appears.

By the way, if I do this:
```sh
npm run build
npm run start
```
The result is the same as that of the development.


# Conclusion
Above all, differences exist between deployment and development at the implement of "@vercel/postgres" package.
My version is `"@vercel/postgres": "^0.4.1",`.
And I know this package is used for Vercel, maybe there are differences between running on my computer and running on Vercel's servers. It may parse the result of the sql sentence in different ways on my side and Vercel's server side. By the way, the deployment's date_time is right.

So, is the difference really exist?
And how do you think of this guess.

# Solution
I solved by modifying the sql sentence.
```
SELECT id, date_time AT ZONE 'UTC' AS utc_date_time FROM meetings;
```
Which returns time_string with time zone mark like `2023-08-25 03:00:00+00`.

And both my computer and Vercel's deploy environment can handle it correctly and the same.

**Comment by kingwingfly** [comment, marked as answer]
https://github.com/vercel/storage/issues/330
